### PR TITLE
D3ASIM-3302: Do not raise exception and force the websocket to discon…

### PR DESCRIPTION
**Pull request by spyrostz**
…nect in case there is an error in the handling of the incoming messages. Corrected logging of incoming trade messages for the aggregator.

### CODE REVIEW

File: d3a_api_client/rest_device.py [hunk 1]
```diff
@@ -166,9 +166,20 @@ def _on_tick(self, message):
...
+        logging.info(f"<-- {message.get('buyer')} BOUGHT {round(message.get('energy'), 4)} kWh "
                      f"at {round(message.get('price'), 2)}/kWh -->")
+
+    def _on_trade(self, message):
+        if "context" not in message:
+            # Device message
+            self._log_trade_info(message)
+        else:
+            # Aggregator message
+            for individual_trade in message["content"]:
```
Comment (faizan2590) [inline d3a_api_client/rest_device.py:180]:
```suggestion
            for "individual_trade" in message["context"]:
```
Not 100% sure though

**Review by faizan2590 — DISMISSED** [review]
One recommendation, other than that LGTM

File: d3a_api_client/rest_device.py [hunk 2]
```diff
@@ -166,9 +166,20 @@ def _on_tick(self, message):
         self.callback_thread.submit(execute_function_util, function=function,
                                     function_name="on_tick")
 
-    def _on_trade(self, message):
-        logging.info(f"<-- {message.get('buyer')} BOUGHT {round(message.get('energy'),4)} kWh "
+    @staticmethod
+    def _log_trade_info(message):
+        logging.info(f"<-- {message.get('buyer')} BOUGHT {round(message.get('energy'), 4)} kWh "
                      f"at {round(message.get('price'), 2)}/kWh -->")
```
Comment (hannesdiedrich) [inline d3a_api_client/rest_device.py:172]: So we are saying that `message.get('energy')` never returns None and skip the None check now?

Comment (hannesdiedrich) [reply d3a_api_client/rest_device.py:172]: Just for clarification, I get the error fixing below.

File: d3a_api_client/websocket_device.py [hunk 3]
```diff
@@ -16,22 +16,26 @@ def __init__(self, rest_client):
...
+                    self.client._on_finish(message)
+                else:
+                    logging.error(f"Received message with unknown event type: {message}")
+            elif "command" in message:
+                self.command_response_buffer.append(message)
 
-        if "event" in message or "command" in message:
-            self.client._on_event_or_response(message)
+            if "event" in message or "command" in message:
+                self.client._on_event_or_response(message)
```
Comment (Hassan754) [inline d3a_api_client/websocket_device.py:35]: Can you please add these 2 lines of code to the Aggregator websocket
```d3a_api_client/aggregator.py:16```

File: d3a_api_client/rest_device.py (hunk 2 above)
Comment (hannesdiedrich) [reply d3a_api_client/rest_device.py:172]: Forget it, yes, you added a new log message that should tell us directly what is happening for better debugging later.

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: d3a_api_client/rest_device.py (hunk 1 above)
Comment (spyrostz) [reply d3a_api_client/rest_device.py:180]: Not really. The message["content"] in the aggregator message contains a list of dicts.

File: d3a_api_client/rest_device.py (hunk 2 above)
Comment (spyrostz) [reply d3a_api_client/rest_device.py:172]: It indeed never returns None, and even if this happens at some point due to a d3a bug, the Exception will be caught and logged by the websocket_device.py -> received_message() and we'll get notified of both the traceback and the incoming message for debugging.

File: d3a_api_client/websocket_device.py (hunk 3 above)
Comment (spyrostz) [reply d3a_api_client/websocket_device.py:35]: Darn I forgot that this method is overridden. Will do that, thanks for pointing it out!

File: d3a_api_client/aggregator.py [hunk 4]
```diff
@@ -14,23 +14,27 @@ def __init__(self, rest_client):
...
+                elif message["event"] == "finish":
+                    self.client._on_finish(message)
+                elif message["event"] == "selected_by_device":
+                    self.client._selected_by_device(message)
+                elif message["event"] == "unselected_by_device":
+                    self.client._unselected_by_device(message)
+                else:
+                    logging.error(f"Received message with unknown event type: {message}")
+            elif "command" in message:
+                self.command_response_buffer.append(message)
```
Comment (Hassan754) [inline d3a_api_client/aggregator.py:34]:
```python
            elif "command" in message:
                self.command_response_buffer.append(message)
        if "event" in message or "command" in message:
            self.client._on_event_or_response(message)
                
```
Do you mind , as a side fix, adding the on_event_or_response callback here, I noticed that due to inheritance this one two needs to be explicitly added

Comment (hannesdiedrich) [reply d3a_api_client/aggregator.py:34]: Yes, I vote for that too, thanks!

Comment (hannesdiedrich) [reply d3a_api_client/aggregator.py:34]: I also promised Sarah that :-)

Comment (spyrostz) [reply d3a_api_client/aggregator.py:34]: Thanks for also catching that ! I'll add it now

**Pull request merged**